### PR TITLE
Stuff

### PR DESCRIPTION
**Pull request by shlomisas**

### CODE REVIEW

File: code/src/basic.js [hunk 1]
```diff
@@ -17,16 +11,16 @@
 (() => {
 
     /**
-     * The task: how you looping `arr` to access the indices and values?
+     * The task: loop thru `arr` to access the indices and values?
```
Comment (imatveev) [inline code/src/basic.js:14]: Redundant question mark

File: code/src/basic.js [hunk 2]
```diff
@@ -17,16 +11,16 @@
...
      */
 
-    let arr = [1, 2, 3];
+    const arr = [1, 2, 3];
 
     /**
-     * The task: how you looping `obj` to access the key and values?
+     * The task: loop thru `obj` to access the key and values?
```
Comment (imatveev) [inline code/src/basic.js:20]: Redundant question mark

File: code/src/functions.js [hunk 3]
```diff
@@ -38,17 +34,23 @@
 })();
 
 (() => {
+    // Assume:
+
+    const foo = (a) => {
+        console.log(a);
```
Comment (imatveev) [inline code/src/functions.js:40]: Let it not use console log, that is an awaited feature to not mix between function and console.log

Comment (shlomisas) [reply code/src/functions.js:40]: amm, not sure i got you, can you explain a bit more?

Comment (imatveev) [reply code/src/functions.js:40]: This task uses `console.log` as an "abstract business logic function to be called in various places of an app" and we also use console log to log the result of the task itself. In 80% cases it leads to confusion. Would be better to have:
```
const foo = (a) => {
        return a;
}
```

File: verbal/backend.md [hunk 4]
```diff
@@ -0,0 +1,30 @@
...
+  * Slow responses
+  * Infinite requests
+  * High CPU
+  * Memory spikes
+  * Memory leaks
+  
+## Queues and messages-brokers
+
+* How a queue works in general? explain the mechanism
+* What's the diff between product/publish and consume/subscribe?
```
Comment (imatveev) [inline verbal/backend.md:15]: Not sure I got this question

File: verbal/backend.md [hunk 5]
```diff
@@ -0,0 +1,30 @@
...
+* How to have multi consumers?
```
Comment (imatveev) [inline verbal/backend.md:16]: The answer is just subscribe to queue multiple times?

File: verbal/backend.md [hunk 6]
```diff
@@ -0,0 +1,30 @@
...
+* How to predict your scaling metrics?
+* How to investigate backend performance issue?
+  * Slow responses
+  * Infinite requests
+  * High CPU
+  * Memory spikes
+  * Memory leaks
+  
+## Queues and messages-brokers
+
```
Comment (imatveev) [inline verbal/backend.md:13]: Would be nice to hear what is the purpose of a queue pattern, which challenges it helps to solve?

File: verbal/backend.md [hunk 7]
```diff
@@ -0,0 +1,30 @@
...
+* How a queue works in general? explain the mechanism
+* What's the diff between product/publish and consume/subscribe?
+* How to have multi consumers?
+
+## NodeJS
+
+* On some endpoint you need to send an email, how you'll do it asynchronously?
```
Comment (imatveev) [inline verbal/backend.md:20]:
```
new Promise((resolve, reject) => {
mailchimp.send('email', (err, result) => {
if (err) {
return rejectt (err);
}
resolve(result)
})
})
```
?

File: code/src/functions.js (hunk 3 above)
Comment (shlomisas) [reply code/src/functions.js]: got you, what do you think about it now?

File: verbal/backend.md (hunk 4 above)
Comment (shlomisas) [reply verbal/backend.md]: sorry, typo, it should be `produce`, rephrased it.

produce/consume - only one consumer catch it as a given moment and can ACK/NACK (typical queue behavior)
publish/subscribe - all subscribers catch it without a ack ability (UDP style)

File: verbal/backend.md (hunk 5 above)
Comment (shlomisas) [reply verbal/backend.md]: sorry, rephrased a bit, the question is about the same message with multiple consumers that all of them need to `ack` they finished what they should do so the main process (etc.) will continue to his next ops.

now it's better?

File: verbal/backend.md (hunk 6 above)
Comment (shlomisas) [reply verbal/backend.md:13]: great point! added.

File: verbal/backend.md (hunk 7 above)
Comment (shlomisas) [reply verbal/backend.md:20]: it's more about `setImmediate` usage with a proper try/catch 

using Promise indeed make an async op (not always, I think I saw some historical libraries implemented Promise synchronously) and even if it's async, in the code you added the main controller flow catchs the error and usually in its `catch` it goes to `next(err)`, which will cause a `ERR_HTTP_HEADERS_SENT` error etc.

**Pull request merged**